test/c/test_sparse_observable.c

```diff
@@ -0,0 +1,650 @@
+
+    // construct the expected observable: 2 * Id
+    QkSparseObservable *expected = qk_obs_zero(100);
+    QkBitTerm bit_terms[] = {};
+    uint32_t indices[] = {};
+    complex double coeff = 2.0;
+    QkSparseTerm term = {coeff, 0, bit_terms, indices, 100};
+    qk_obs_add_term(expected, &term);
+
+    bool is_equal = qk_obs_equal(expected, simplified);
+
+    qk_obs_free(obs);
+    qk_obs_free(right);
+    qk_obs_free(left);
+    qk_obs_free(simplified);
+    qk_obs_free(expected);
+
+    if (!is_equal) {
+        return EqualityError;
+    }
+
+    return Ok;
+}
+
+/**
+ * Test getting the number of terms in an observable.
+ */
+int test_num_terms() {
+    int result = Ok;
+    uint64_t num_terms;
+
+    QkSparseObservable *zero = qk_obs_zero(100);
+    num_terms = qk_obs_num_terms(zero);
+    if (num_terms != 0) {
+        result = EqualityError;
+    }
+    qk_obs_free(zero);
+
+    QkSparseObservable *iden = qk_obs_identity(100);
+    num_terms = qk_obs_num_terms(iden);
+    if (num_terms != 1) {
+        result = EqualityError;
+    }
+    qk_obs_free(iden);
+
+    return result;
+}
+
+/**
+ * Test getting the number of qubits in an observable.
+ */
+int test_num_qubits() {
+    int result = Ok;
+    uint32_t num_qubits;
+
+    QkSparseObservable *obs = qk_obs_zero(1);
+    num_qubits = qk_obs_num_qubits(obs);
+    if (num_qubits != 1) {
+        result = EqualityError;
+    }
+    qk_obs_free(obs);
+
+    QkSparseObservable *obs100 = qk_obs_zero(100);
+    num_qubits = qk_obs_num_qubits(obs100);
+    if (num_qubits != 100) {
+        result = EqualityError;
+    }
+    qk_obs_free(obs100);
+
+    return result;
+}
+
+/**
+ * Test adding an individual term to an observable.
+ */
+int test_custom_build() {
+    u_int32_t num_qubits = 100;
+    QkSparseObservable *obs = qk_obs_zero(num_qubits);
+
+    complex double coeff = 1;
+    QkBitTerm bit_terms[3] = {QkBitTerm_X, QkBitTerm_Y, QkBitTerm_Z};
+    uint32_t indices[3] = {0, 1, 2};
+    QkSparseTerm term = {coeff, 3, bit_terms, indices, num_qubits};
+
+    qk_obs_add_term(obs, &term);
+    qk_obs_add_term(obs, &term);
+
+    double tol = 1e-6;
+    QkSparseObservable *simplified = qk_obs_canonicalize(obs, tol);
+
+    uint64_t num_terms = qk_obs_num_terms(obs);
+    uint64_t num_terms_simplified = qk_obs_num_terms(simplified);
+
+    qk_obs_free(obs);
+    qk_obs_free(simplified);
+
+    if (num_terms != 2 || num_terms_simplified != 1) {
+        return EqualityError;
+    }
+    return Ok;
+}
+
+/**
+ * Test getting the terms in an observable.
+ */
+int test_term() {
+    uint32_t num_qubits = 100;
+    QkSparseObservable *obs = qk_obs_identity(num_qubits);
+
+    QkBitTerm bit_terms[3] = {QkBitTerm_X, QkBitTerm_Y, QkBitTerm_Z};
+    uint32_t qubits[3] = {0, 1, 2};
+    complex double coeff = 1 + I;
+
+    QkSparseTerm term = {coeff, 3, bit_terms, qubits, num_qubits};
+    int err = qk_obs_add_term(obs, &term);
+
+    if (err != 0) {
+        return RuntimeError;
+    }
```

Comment:
This is one place where we early return without cleaning up, so the `goto cleanup;` pattern I'd suggested earlier _might_ be appropriate.

That said: do we know how many separate processes CTest runs? Does it do every function in a separate process?  If it does: it's maybe not really any trouble if we "leak" memory, since it's all reclaimed on process termination anyway, and the memory use of the test is clearly well bounded.  That said, it's quite conceivable we might want to run the test suite under very address / memory / behaviour sanitisers in the future, which would fail us if we leak.

Reply:
oh yeah, I'll add the missing free here -- the `cleanup` sounds like a good pattern though we'd have to check what to do in places where we allocate more objects during the test, then we don't always have to free the same variables

Reply:
Updated in all occurences in 624c3dd